Add Nix/NixOS dev-shell

qtdyn opened the pull request:
This PR adds a Nix/NixOS development-shell (`shell.nix`), which is based on the upstream [nixpkgs](https://github.com/NixOS/nixpkgs/blob/c5d4d458115263ec3ee0efd974e6aeb2f787a0ed/pkgs/by-name/ze/zed-editor/package.nix), as well as its corresponding `flake.nix` file.

To use it, run either the `nix-shell` command (uses the `shell.nix` file), or the newer but experimental `nix develop` command (uses `flake.nix`)

~~This has not been tested on macOS, tho preliminary code is there to try and support it, feel free to report any issues.~~ Zed unfortunately doesn't build on nix-darwin (see https://github.com/NixOS/nixpkgs/issues/320084), so this PR doesn't aim to add darwin support.

---

Release Notes:

- N/A

qtdyn committed:
Add nix development shell w/ flake support

Signed-off-by: xtrm <oss@xtrm.me>

xtrm-en commented:
@cla-bot check

mrnugget commented on .envrc:
```diff
@@ -0,0 +1 @@
+use flake
```
I think adding a `flake.nix` file is fine, but prescribing it for all direnv users I'm skeptical about.

mrnugget replied:
I'd remove the `.envrc` file and add it to `.gitignore` too. That way people can choose to use it or not.

maan2003 replied:
standard practice is: change it to `.envrc.sample` and ask users to `mv  .envrc.sample .envrc` in setup instruction for nix (and with .gitignore)

xtrm-en replied:
I'd opt with just removing it since it's a single line, and providing a simple will force to pick between flake and non-flake usage.

niklaskorz commented:
FYI, Zed does not build in Nix on Darwin yet, so it's probably best to just remove the darwin-specific parts from this PR to not give a false impression. See https://github.com/NixOS/nixpkgs/issues/320084

Building Zed on macOS is generally easy anyway (outside of Nix) and I don't think there's much benefit of the dev shell there, while for NixOS it of course saves a lot of trouble.

niklaskorz commented on shell.nix:
```diff
@@ -0,0 +1,64 @@
+pkgs.mkShell {
+  nativeBuildInputs =
+    with pkgs;
+    [
```
Add `clang` and `mold` here. These packages are required because of https://github.com/zed-industries/zed/blob/main/.cargo/config.toml

niklaskorz commented:
Zed has a runtime dependency on libstdc++, currently `cargo run` with this nix-shell fails:

```
target/release/zed: error while loading shared libraries: libstdc++.so.6: cannot open shared object file: No such file or directory
```

Also, we need to have the other non-Rust dependencies in our library path if we don't want to run patchelf on the binaries cargo produces.
We also need to add vulkan-loader to the library path so it can be found at runtime.
This can all be fixed by adding a `LD_LIBRARY_PATH` to the shell.nix:

```nix
  LD_LIBRARY_PATH = with pkgs; lib.makeLibraryPath (buildInputs ++ [
    stdenv.cc.cc.lib
    vulkan-loader
  ]);
```

(requires `pkgs.mkShell rec {...` so we can access `buildInputs` here)

mrnugget commented:
Thanks for the comments @niklaskorz, very much appreciated :)

qtdyn committed:
Remove `.envrc` file

Signed-off-by: xtrm <oss@xtrm.me>

qtdyn committed:
Implement changes requested

This removes initial shell support for Darwin as nix-darwin isn't currently able to build zed (see https://github.com/NixOS/nixpkgs/issues/320084)

This also fixes running the resulting cargo binaries as there were missing dependencies (@niklaskorz)

Co-authored-by: Niklas Korz <niklas@niklaskorz.de>
Signed-off-by: xtrm <oss@xtrm.me>

qtdyn committed:
Throw when running on Darwin, update flake.lock

Signed-off-by: xtrm <oss@xtrm.me>

mrnugget commented:
Let's try this! Seems like a non-intrusive addition that we can always change/revert.

JJ-Atkinson commented:
Just for some feedback here, this is great. I was looking to add a dead simple flake nix for my local dev, but having it in repo already is very nice.
